File: algorithms/1_easy/F/00387_first_unique_character_in_a_string/solution.c
#define INF SIZE_MAX

enum state {
    NOT_SEEN,
    ONCE_SEEN,
    MULTI_SEEN,
};

int firstUniqChar(const char *restrict s) {
    enum state states[256] = { 0 };
    size_t first_seen[256] = { 0 };

    for (size_t i = 0; s[i] != '\0'; i += 1) {
        const unsigned char ch = (unsigned char)s[i];

        if (states[ch] == NOT_SEEN) {
            states[ch] = ONCE_SEEN;
            first_seen[ch] = i;
        } else {
            states[ch] = MULTI_SEEN;
        }
    }

    size_t res = INF;

    for (size_t i = 0; i < 256; i += 1) {
        if (states[i] == ONCE_SEEN) {
            res = MIN(res, first_seen[i]);
        }
    }

    if (res == INF) {
        return -1;
    }

    return (int)res;
}

File: algorithms/1_easy/F/00387_first_unique_character_in_a_string/solution.rs
#[derive(Clone, Copy, Eq, PartialEq)]
enum State {
    NotSeen,
    OnceSeen(usize),
    MultiSeen,
}

impl Solution {
    pub fn first_uniq_char(s: String) -> i32 {
        let mut states = [State::NotSeen; 256];

        for (i, &byte) in s.as_bytes().into_iter().enumerate() {
            if states[byte as usize] == State::NotSeen {
                states[byte as usize] = State::OnceSeen(i);
            } else {
                states[byte as usize] = State::MultiSeen;
            }
        }

        let mut res: Option<usize> = None;

        for state in states {
            if let State::OnceSeen(idx) = state {
                res = Some(res.map_or(idx, |other_idx| other_idx.min(idx)));
            }
        }

        match res {
            None => -1,
            Some(idx) => idx.try_into().unwrap(),
        }
    }
}
